### ISSUE TITLE
Overfitting in CEBRA Supervised Embeddings

### ISSUE THREAD

**Issue by jgraving** [AdaptiveMotorControlLab/CEBRA#226]
Hi @stes and @MMathisLab et al.,

 I've been testing CEBRA supervised embeddings and observed that even when the neural data are shuffled, the resulting embeddings remain highly consistent with those for the original (unshuffled) data—with high decoding accuracy and low InfoNCE loss. This suggests that the model is overfitting to the supervised labels, and high consistency and decoding performance is the null expectation for any sufficiently flexible encoder network. Naturally, this calls into question the soundness of the method as well as the validity of any scientific inferences drawn from the use of such a method.

- **Method:**  
Experiments were conducted for example rat and monkey data using the provided code for CEBRA supervised embeddings. Embeddings for the original (unshuffled) and shuffled (randomly permuted) neural data were compared. Supervised labels were left intact (unshuffled).

- **Code & Models:**  
Code, figures, and saved models for the experiments are available at https://github.com/jgraving/cebra_shuffle_overfit

- **Observations:**  
    - When the neural data are shuffled, the embeddings remain highly consistent with those for the original (unshuffled) data.
    - Decoding accuracy remains high.
    - InfoNCE loss remains low.

**Concerns**

- **Overfitting:**  
  Experiments strongly indicate that the encoder network for the neural data is readily overfitting to the supervised labels.

- **Lack of Proper Validation:**  
  The software package doesn't provide any obvious functionality for proper validation and testing of generalization for supervised embeddings, and it doesn’t appear that this issue is discussed or addressed in the paper.

- **Documentation Issues:**  
  The documentation encourages users to increase the network size ad libitum for a "better embedding" without any mention of pitfalls. This effectively suggests that users should overfit embeddings to their data.
https://cebra.ai/docs/api/sklearn/cebra.html#cebra.CEBRA.num_hidden_units 
```
num_hidden_units (int): The number of dimensions to use within the neural network model.
Higher numbers slow down training, but make the model more expressive and can result in a better embedding.
Especially if you find that the embeddings are not consistent across runs, increase :py:attr:~.num_hidden_units and :py:attr:~.output_dimension 
to increase the model size and output dimensionality.
|Default:| 32.
```

- **Impact:**  
These limitations raise serious concerns about the central claims and soundness of the work, especially given the high-impact venue where the paper was published and the seemingly widespread adoption of the method.

**Proposed Next Steps**

1. **Paper Correction and Software Update:**  
 The paper should be corrected and the software package updated to address these limitations—or retracted if the issues cannot be adequately resolved.

2. **User Awareness:**  
You as the maintainers/authors should use github and your social media channels to help alert users to this issue so that they can make an informed decision about whether or not to use the method in its current form and avoid potentially wasting time and resources applying a seemingly unreliable method.

I invite you as maintainers, as well as others contributors and users, to review these findings as well as investigate this problem yourselves. This issue aims to ensure that you and other users are aware of this critical limitation and that necessary actions are taken to address it.

Best regards,  
Jake Graving

**Comment by MMathisLab** [comment]
Hi @jgraving, thanks for your post. **The issue you observe can occur when on limited data, too large of models get heavily overtrained, and no train/validation splits or sanity checks on the training curves are performed to uncover such an issue.** This is true for any deep learning-based approach. 

We therefore want to immediately push back that our paper should be retracted or corrected; we never make any claims that it's impossible to overfit, and below we provide both intuition about why your experiments are not what we show in our paper, and that your experiments do not invalidate our claims (i.e., plainly speaking, we show results with reasonable and justifiable model parameters and training). 

## Shuffle controls: 

In fact, it is entirely expected that the distribution of the labels shape the embedding (see [Proposition 7, Supplementary Note 2](https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-023-06031-6/MediaObjects/41586_2023_6031_MOESM1_ESM.pdf)). In [Figure 2c](https://www.nature.com/articles/s41586-023-06031-6#Fig2) we therefore show shuffle controls demonstrating that if **labels are shuffled**, it is not possible to fit them.

In your example, you shuffle the neural data. If the model has sufficient capacity and the data is limited, it can fit a “lookup table” from input data to the output embedding to match the label distribution (because this is intact and still has structure), which is what you observe in your example.


## Overfitting:

For hypothesis testing applications of CEBRA, **it is crucial to closely monitor the loss curves**. A typical signature of overfitting is a high loss in the beginning (around log N, typically a bit smaller due to numerical effects), and then after many iterations a drop when the model enters the overfitting regime. This is exactly what is observable in your dataset. 

Below we [plot your provided achilles embeddings](https://colab.research.google.com/drive/1tUJTCYyxVK53LBtr3VdMcWLIL-gjh3D0?usp=sharing) (neural shuffled and non-shuffled) and the expected behavior can be clearly seen: (a) smaller models have the overfitting point much later in training, larger models overfit earlier (b) overfitting leads to a sudden drop very late in training:

![Image](https://github.com/user-attachments/assets/6a297b74-8b74-4c89-aaba-0bafc4f14a4f)

For ease of reference, in our publication we used hidden units = 32. If we use your neural shuffling scheme, at a reasonable point in training, we see a clear difference between shuffle and non-shuffled, as expected:

<img src="https://github.com/user-attachments/assets/67c953ce-1a63-4e62-92e6-db9fec0c84bb" width="40%" /> 
<img src="https://github.com/user-attachments/assets/a7c9c49f-7ae1-4b1f-85d5-90af14269c9f" width="43%" />

## Consistency 

In your example, visualizing the consistency across multiple steps of training and/or tracking consistency on both train and validation set would be important, if in doubt.

For example, here is the four rats from the hippocampus dataset with your neural-shuffle, time-only, and non-shuffled; at 5K steps this is the result:

![Image](https://github.com/user-attachments/assets/57dd796f-622c-416b-a622-de16015d6f84)

![Image](https://github.com/user-attachments/assets/620db83d-d6e5-46f3-b278-023faea39f2f)

## Hyperparameter tuning:

Contrary to your claim, we do provide built-in tooling for easily [plotting loss curves](https://cebra.ai/docs/api/integrations/matplotlib.html#cebra.integrations.matplotlib.plot_loss) and [performing train/validation splits of data before settling on parameters](https://cebra.ai/docs/usage.html#quick-start-scikit-learn-api-example).

In our user guide we suggest performing a train/validation split to be sure that, in addition to shuffle controls, one’s parameters don’t drive overfitting (the essence of a train/val). To note, just like in UMAP or tSNE where there is not a train vs. test split typically, one doesn’t necessarily need to only use validation for comparing models. Users should just choose parameters appropriately before training on the full data.

**For decoding applications, it is crucial to perform a train/validation split.** 






## Updates: let’s be constructive


### Updated demo

While **our main hippocampus demo notebook already included how to do the train/validation**, we added a new section at the top to highlight this more [PR here](https://github.com/AdaptiveMotorControlLab/CEBRA-demos/pull/19). We see this as a step to validate parameters as we say in our user guide. As stated above,  just like in UMAP or tSNE where there is not a train vs. test split typically, one doesn’t necessarily need to only use validation for comparing models. Users should just choose parameters appropriately before training on the full data.

### Note in User Guide:

In our [9 Steps to work with CEBRA](https://cebra.ai/docs/usage.html#quick-start-scikit-learn-api-example), the Step 3, already in our docs clearly highlights to users they can use a train/validation set:

```python
 # 3. Split data and labels
     (
         train_data,
         valid_data,
         train_discrete_label,
         valid_discrete_label,
         train_continuous_label,
         valid_continuous_label,
     ) = train_test_split(neural_data,
                         discrete_label,
                         continuous_label,
                         test_size=0.3)
```

But, we added a [PR](https://github.com/AdaptiveMotorControlLab/CEBRA/pull/229)  to update the docs to educate users about spotting this issue when applying on their datasets earlier on:

```
#. To note, you should still be mindful of performing train/validation splits and shuffle controls to avoid `overfitting <https://developers.google.com/machine-learning/crash-course/overfitting/overfitting>`_.
```



Best Regards, 
Mackenzie @mmathislab and Steffen @stes

**Comment by jgraving** [comment]
Hi @stes and @MMathisLab,

Thanks for the explanation and updates. I realize now my earlier suggestion to retract or correct the paper was an overstatement. My intent was simply to highlight a potentially important issue, which you have addressed, so I will close the issue.

Best,
Jake

**Comment by MMathisLab** [comment]
Hi @jgraving,

before digging deeper into this, the **numbers you are sharing here are strictly worse** than the ones in our demo notebook for decoding (https://cebra.ai/docs/demo_notebooks/Demo_decoding.html). Could you check this out, and let us know if that already contributes to solving your concern?

Please also note that consistency alone is not the metric to rely on. The loss, as we already pointed out, clearly matters here: 20K iterations overfits rat 1 in your own prior demo, and we do not use 20K in our paper. In Supplementary Fig. 7 (https://cebra.ai/docs/cebra-figures/figures/ExtendedDataFigure7.html), for instance, where we compare single and multi session embedding, we compare consistency at a fixed decoding performance:

![Image](https://github.com/user-attachments/assets/c9fe3759-2f64-4f84-b4b7-54db5e09b197)

The results in this notebook (**on the held out test set**) clearly demonstrate that decoding is possible on a held-out portion of the dataset (all of our reported decoding is on held out data). A comparison between InfoNCE and the decoding performance is also provided in the last cell of [the notebook](https://cebra.ai/docs/demo_notebooks/Demo_decoding.html):

![Image](https://github.com/user-attachments/assets/64b0ec69-35a7-45f7-a50b-0606c24f3381)

I would also note, infoNCE is the proper metric to use given that is our loss, and that is the only metric we used in our paper. We recently included GoF  in the codebase to help ease interpretation of model fit, but again, unrelated to our paper (and GoF on the small validation data is noisy and not interpretable and we never suggest looking at this metric on validation unless with very large datasets).

**Comment by jgraving** [comment]
Hi @MMathisLab, 

Yes I explored a range of training steps as you can clearly see in the plots. Results change very little after a couple thousand training steps. Your example uses 5000, which is included.

 It seems your results are using median error not mean, which might explain the difference. Median effecitvely ignores the tail of the distribution, so higher magnitude errors are weighted less heavily.

And you argue goodness of fit is noisy for small datasets? But it is a bijection of infonce simply shifted by an additive constant https://github.com/AdaptiveMotorControlLab/CEBRA/blob/823c9ca8378beca3e029988fd85a2932c9e3c405/cebra/integrations/sklearn/metrics.py#L251 if goodness of fit is noisy then so is infonce, and so your argument is then that neither should be used for detecting overfitting?

**Comment by jgraving** [comment]
Hi @MMathisLab ,

The median error is indeed closer to your published/example results:
<img src="https://raw.githubusercontent.com/jgraving/cebra_shuffle_overfit/main/cebra_figures/hippocampus_pos_decoding_median_error_train_test.png" alt="Hippocampus Decoding Median Error" width="600">

And, yes it works well only on part of the test set, as you mentioned:
<img src="https://raw.githubusercontent.com/jgraving/cebra_shuffle_overfit/main/cebra_figures/hippocampus_pos_decoding_consistency_test.png" alt="Hippocampus Decoding Consistency Test" width="600">

It doesn't seem that infonce or goodness of fit are all that useful for detecting overfitting, especially when the task is decoding. But then brute forcing the hyperparameter search to find good decoding generalization is inefficient and requires a lot of effort (and compute) on the part of the user. Perhaps something to consider for future updates.

Either way, my results now match yours sufficiently well. Not quite the performance I expected, but it is what it is. So I will go ahead and close the issue. 

Thanks for your time and efforts.

Best,
Jake

**Comment by jgraving** [comment]
Thanks, that makes more sense. I was using the train test split you posted in your previous reply, but I guess this was just meant as a generic example.

> ### Note in User Guide:
> In our [9 Steps to work with CEBRA](https://cebra.ai/docs/usage.html#quick-start-scikit-learn-api-example), the Step 3, already in our docs clearly highlights to users they can use a train/validation set:
> 
>  # 3. Split data and labels
>      (
>          train_data,
>          valid_data,
>          train_discrete_label,
>          valid_discrete_label,
>          train_continuous_label,
>          valid_continuous_label,
>      ) = train_test_split(neural_data,
>                          discrete_label,
>                          continuous_label,
>                          test_size=0.3)
Also in your quick start guide you import train test split from sklearn but then don't use it and split the last contiguous indices instead. 

Seems my other error was decoding position only from a position direction embedding. Also makes sense to decode the same labels used for training and then calculate the error for position only.

**Comment by jgraving** [comment]
Just to confirm, the example you sent previously was using sklearn train test split so that's what I used. Seems you've now fixed this but didn't mention it...

https://github.com/AdaptiveMotorControlLab/CEBRA-demos/blob/cfd35addabc3be68a49252a2c62557b54abc1ee8/Demo_hippocampus.ipynb

**Comment by MMathisLab** [comment]
@jgraving actually no; the decoding demo notebook always has had the correct split (as I edited your notebook to have), and is what I linked above and asked you to use/investigate: https://github.com/AdaptiveMotorControlLab/CEBRA-demos 

![Image](https://github.com/user-attachments/assets/dc62b856-e68a-4893-800a-8bbc1cedc792)

The usage doc indeed had a general `sklearn` split demo (which can be fine for some data/models), and on Saturday as part of some minor refactoring to take in your feedback, which I pointed out in my "lets be constructive" section,  for the **WIP** notebook it had it ... the current merged and displayed demo does not have this.

Can I ask, what is the goal here? We have civilly engaged with your questions, spent a lot of time on this to address your concerns, took your feedback to update any clarifications, and spent hours going though and finding errors in your demo that is attempting to "retract our paper"... at this point, I respectfully ask we end here.

**Comment by jgraving** [comment]
@MMathisLab Fair enough. Thanks for your patience and further clarifications. I was indeed looking at the WIP notebook. I only mentioned it because you said I deviated from your examples when I was in fact using the code you specifically pointed me to for how to perform validation:

![Image](https://github.com/user-attachments/assets/f31a8624-7be6-42ca-82ed-c4aaf57c343c)

To summarize, my initial concerns arose when I ran a basic shuffle control on some neural data with the default parameters, which resulted in overfitting that became more pronounced with larger model sizes and extended training. I also couldn’t locate code for detecting or preventing overfitting—for example, using a validation set for early stopping—nor did I find any discussion of this failure mode in the paper. Consequently, I attempted to replicate the results using your examples to bring the issue to your attention. You then directed me to the validation code and suggested that I perform a parameter search to find settings that work well on the validation set. However, when I wasn’t able to replicate the expected decoding performance using the provided code, I reopened the issue.

Clearly I should have reviewed the documentation more thoroughly before opening the issue, and I regret having suggested retraction or correction—that was an unnecessary overstatement. I appreciate the time you and @stes have dedicated to addressing my questions and updating the documentation. Thanks again for your guidance.
